README.md
<p align="left"><img src="images/logo/horizontal.png" alt="Sounder" height="120px"></p>

# Py_MusicPlayer
Python mp3 music player.

## General Features:
| Goals | Status |
| --- | --- |
| adding mp3 file support | ☑ |
| adding xm file support | ☑ |
| adding wav file supports | ☑ |
| adding ogg file support | ☑ |  
| adding changelog to Sounder | ☑ | 
| adding updater to Sounder | ☑ | 
## Icons
https://icons8.com/

## Authors
Main programmer, designer: Mateusz Perczak
<br></br>
Logo designer: reallinfo

## Python version:
3.6.8

## Used Libraries:

+ os
+ time
+ threading
+ json
+ pygame
+ tkinter
+ PIL
+ io
+ mutagen
+ typing
+ logging
+ requests
+ zipfile
+ random
+ psutil

## Some pictures of application
![Picture of playing music in application](https://github.com/losek1/Sounder3/blob/master/images/NEW0.PNG)
![Picture of settings in application](https://github.com/losek1/Sounder3/blob/master/images/NEW1.PNG)
![Picture without any playing song](https://github.com/losek1/Sounder3/blob/master/images/NEW2.PNG)
![Picture of dark mode](https://github.com/losek1/Sounder3/blob/master/images/NEW3.PNG)


src/changelog.txt
~Some code optimizations.
Note: This is final update :(


src/test.py
from tkinter import *
from PIL import Image, ImageSequence, ImageTk
from threading import Thread
from tkinter import ttk
from typing import ClassVar, List
from time import sleep
main_window: ClassVar = Tk()
main_window.title("ANIMATION")
main_window.geometry("300x300")
main_window.iconbitmap("icon.ico")
# main_window.resizable(width=FALSE, height=FALSE)
main_window.configure(background="#fff")
updater_theme = ttk.Style()
updater_theme.theme_use('clam')
updater_theme.configure("TLabel", background='#fff', foreground='#000', border='0')
img_frames: List = []


def load_img():
    global img_frames
    img: ClassVar = Image.open("download_light.gif")
    for frame in ImageSequence.Iterator(img):
        img_frames.append(ImageTk.PhotoImage(frame.copy().convert('RGBA').resize((200, 200))))
    while True:
        for frame in img_frames:
            image_label.configure(image=frame)
            sleep(0.015)


Thread(target=load_img, daemon=True).start()
image_label: ClassVar = ttk.Label(main_window, image=None)
image_label.place(relx=0.5, rely=0.5, anchor="center")
main_window.mainloop()

src/entry.py
from tkinter import *
from tkinter import ttk
from typing import ClassVar, Dict, List
from tkinter.filedialog import askdirectory
from os import chdir, listdir
from re import findall
from threading import Thread

root: ClassVar = Tk()
root.geometry("400x700")
root.title("Entry test")
root.configure(background="#fff")

root_theme = ttk.Style()
root_theme.theme_use('clam')
items: List = []
search_list: ClassVar = []


def add_files() -> None:
    global items
    target_dir: str = "C:/Users/Mateusz/Music"
    if bool(target_dir):
        chdir(target_dir)
        for i, file in enumerate(listdir(target_dir)):
            if file.endswith(".mp3"):
                items.append(file)
                list_box.insert(i + 1, file)


def search(e=None) -> None:
    global items, search_list
    if bool(search_box.get()):
        list_box.select_clear(0, END)
        list_box.delete(0, END)
        search_list = []
        for item in items:
            if bool(findall(search_box.get().lower(), item.lower())):
                list_box.insert(len(items), item)
                search_list.append(item)
        if bool(search_list):
            seperator.configure(background='#00E676')
        else:
            seperator.configure(background='#B31412')
    else:
        list_box.select_clear(0, END)
        list_box.delete(0, END)
        for item in items:
            list_box.insert(len(items), item)
        seperator.configure(background='#000')


def list_box_select(e=None) -> None:
    global search_list, items
    if bool(search_box.get()) and bool(search_list):
        print(search_list[list_box.curselection()[0]])
    elif bool(items) and bool(list_box.curselection()):
        print(items[list_box.curselection()[0]])


search_frame: ClassVar = Frame(root)
search_frame.configure(background="#fff")
seperator: ClassVar = Frame(search_frame)
seperator.configure(background="#000")
seperator.place(relx=0.5, rely=0.6, relwidth=0.99, relheight=0.4, anchor='n')
search_box: ClassVar = Entry(search_frame, font=('Bahnschrift', 18), exportselection=1, relief='flat',
                             selectbackground="#000", selectforeground="#fff", )

search_box.place(relx=0.5, rely=0.06, relwidth=0.99, relheight=0.86, anchor='n')

search_frame.place(relx=0.5, rely=0.01, relwidth=0.99, relheight=0.1, anchor='n')

list_box: ClassVar = Listbox(root, font=('Bahnschrift', 11), relief="flat", exportselection=0, cursor="hand2"
                             , bd=0, activestyle="none", takefocus=False, selectmode="SINGLE",
                             highlightthickness=0, selectbackground="#B31412", foreground='#000', background='#fff',)
list_box.place(relx=0.5, rely=0.12, relwidth=0.99, relheight=0.87, anchor='n')

search_box.bind("<KeyRelease>", search)
list_box.bind("<<ListboxSelect>>", list_box_select)
Thread(target=add_files, daemon=True).start()
root.mainloop()


src/Updater.py
try:
    from os import getcwd, startfile, rename, remove
    from os.path import dirname, isfile, basename
    from threading import Thread
    from json import load
    from tkinter import Tk, PhotoImage, sys, Frame
    from tkinter import ttk
    from typing import ClassVar, Dict, List
    from requests import get
    from zipfile import ZipFile
    import logging
    from io import BytesIO
    from time import sleep
    from psutil import process_iter
    from PIL import Image, ImageSequence, ImageTk
except ImportError as e:
    sys.exit(e)

# dir
sounder_dir: str = getcwd()
# sounder_dir: str = dirname(sys.executable)
# log
logging.basicConfig(filename=f"{sounder_dir}\\errors.log", level=logging.ERROR)
# window setup
updater_window: ClassVar = Tk()
updater_window.withdraw()
updater_window.geometry(f"375x100+{int(updater_window.winfo_x() + ((updater_window.winfo_screenwidth() - 375) / 2))}"
                        f"+{int(updater_window.winfo_y() +((updater_window.winfo_screenheight() - 100) / 2))}")
updater_window.title("Sounder updater")
try:
    updater_window.iconbitmap(f"{sounder_dir}\\icon.ico")
except:
    sys.exit(1)
updater_window.resizable(width=False, height=False)
updater_window.configure(background="#fff")
# theme
updater_theme = ttk.Style()
updater_theme.theme_use('clam')
updater_theme.configure("TLabel", background='#fff', foreground='#000', border='0')
updater_theme.configure("Horizontal.TProgressbar", foreground='#fff', background='#000', lightcolor='#fff'
                        , darkcolor='#fff', bordercolor='#fff', troughcolor='#fff')
updater_theme.configure("TButton", relief="flat", background='#000', font=('Bahnschrift', 10), foreground='#fff')
updater_theme.map("TButton", background=[('pressed', '!disabled', '#000'), ('active', '#111')])
# variables
config: Dict = {}
server_version: str
package = b''

# functions


def dump(err_obj: ClassVar) -> None:
    error_reason_label.configure(text="Error: " + logging.getLevelName(err_obj))
    show(error_frame)
    logging.error(err_obj, exc_info=True)


def show(window) -> bool:
    try:
        window.tkraise()
        return True
    except Exception as e:
        dump(e)
        return False


def load_config() -> bool:
    global config
    if isfile('cfg.json'):
        try:
            with open('cfg.json', 'r') as data:
                config = load(data)
            return True
        except:
            return False
    else:
        return False


def close() -> None:
    for widget in updater_window.winfo_children():
        widget.destroy()
    updater_window.destroy()
    sys.exit(0)


def change_mode(mode: str) -> None:
    try:
        update_progress.stop()
        update_progress["value"] = 0
        update_progress["maximum"] = 100
        if mode == "determinate":
            update_progress.configure(mode="determinate")
        elif mode == "indeterminate":
            update_progress.configure(mode="indeterminate")
            update_progress.start(4)
    except Exception as e:
        dump(e)


def update() -> bool:
    global server_version, package, sounder_dir
    chunk_size: int = 8192
    change_mode("determinate")
    Thread(target=load_img, daemon=True).start()
    show(checking_frame)
    try:
        bytes_downloaded: float = 0
        server_zip = get(f"https://github.com/losek1/Sounder3/releases/download/v{server_version}/package.zip"
                         , stream=True)
        if server_zip.status_code == 200:
            update_progress["maximum"] = int(server_zip.headers.get('Content-Length'))
            show(update_frame)
            for chunk in server_zip.iter_content(chunk_size=chunk_size):
                if chunk:
                    package += chunk
                    bytes_downloaded += chunk_size
                    update_progress["value"] = bytes_downloaded
                    update_data.configure(text=f"{round(bytes_downloaded / 1000000, 1)}MB / {round(int(server_zip.headers.get('Content-Length')) / 1000000, 1)}MB")
            for process in process_iter():
                if process.name() == "Sounder3.exe":
                    process.kill()
            change_mode("indeterminate")
            update_label.configure(text="Installing updates ...")
            with ZipFile(BytesIO(package)) as zip_file:
                for file in zip_file.namelist():
                    update_data.configure(text=f"{zip_file.namelist().index(file)} / {len(zip_file.namelist())}")
                    if file == "Updater.exe" or file == "errors.log":
                        continue
                    try:
                        zip_file.extract(file, sounder_dir)
                    except Exception as error_obj:
                        logging.error(error_obj, exc_info=True)
            show(finish_frame)
            sleep(3)
            open_app()
            close()
        else:
            raise Exception("Cannot contact GitHub servers")
    except Exception as e:
        dump(e)
        return True


def check_updates() -> bool:
    global config, server_version
    try:
        server_version = get(
            "https://raw.githubusercontent.com/losek1/Sounder3/master/updates/version.txt").text.strip()
        if int(config["version"].replace(".", "")) < int(server_version.replace(".", "")):
            return True
        else:
            return False
    except Exception as e:
        dump(e)


def open_app() -> None:
    try:
        if isfile("Sounder3.exe"):
            startfile("Sounder3.exe")
    except Exception as e:
        dump(e)


def update_task() -> None:
    Thread(target=update, daemon=True).start()


def self_upgrade() -> bool:
    try:
        if basename(sys.argv[0]) == "New-Updater.exe":
            if isfile("Updater.exe"):
                remove("Updater.exe")
            rename(sys.argv[0], "Updater.exe")
            return False
        elif isfile("New-Updater.exe"):
            startfile("New-Updater.exe")
            return True
        else:
            return False
    except Exception as e:
        dump(e)


def show_gui() -> None:
    show(checking_frame)
    updater_window.deiconify()
    updater_window.lift()
    updater_window.focus_force()


def init() -> None:
    if self_upgrade():
        close()
    elif load_config():
        show_gui()
        if check_updates():
            update_task()
        else:
            show(choice_frame)


def load_img() -> None:
    try:
        img_frames: List = []
        download_img: ClassVar = Image.open("download_light.gif")
        for frame in ImageSequence.Iterator(download_img):
            img_frames.append(ImageTk.PhotoImage(frame.copy().convert('RGBA').resize((48, 48))))
        if len(img_frames) > 1:
            while True:
                for frame in img_frames:
                    update_img_label.configure(image=frame)
                    sleep(0.02)
        else:
            update_img_label.configure(image=img_frames)
    except Exception as e:
        dump(e)


# frames
# error frame
error_frame: ClassVar = Frame(updater_window)
error_frame.configure(background="#fff")
error_reason_label: ClassVar = ttk.Label(error_frame, text="Error:", anchor='center', font='Bahnschrift 11')
error_exit_button: ClassVar = ttk.Button(error_frame, text="EXIT", cursor="hand2", takefocus=False, command=close)
error_reason_label.place(relx=0.5, rely=0, relheight=0.58, anchor="n")
error_exit_button.place(relx=0.5, rely=0.6, relwidth=0.23, anchor="n")
error_frame.place(relx=0, rely=0, relwidth=1, relheight=1)
# end
# update frame
update_frame: ClassVar = Frame(updater_window)
update_frame.configure(background="#fff")
update_img_label: ClassVar = ttk.Label(update_frame, image=None, anchor='center')
update_label: ClassVar = ttk.Label(update_frame, text="Downloading updates ...", font='Bahnschrift 11')
update_data: ClassVar = ttk.Label(update_frame, text="-- MB / --MB", font='Bahnschrift 10', anchor='center')
update_progress: ClassVar = ttk.Progressbar(update_frame, orient='horizontal')
update_img_label.place(relx=0, rely=0, relwidth=0.25, relheight=1)
update_label.place(relx=0.25, rely=0.15, relwidth=0.45, relheight=0.25)
update_data.place(relx=0.70, rely=0.15, relwidth=0.30, relheight=0.25)
update_progress.place(relx=0.25, rely=0.6, relwidth=0.75, relheight=0.25)
update_frame.place(relx=0, rely=0, relwidth=1, relheight=1)
# end
# check frame
checking_frame: ClassVar = Frame(updater_window)
checking_frame.configure(background="#fff")
checking_label: ClassVar = ttk.Label(checking_frame, text="Verifying\n"
                                                          ". . .", font='Bahnschrift 16', anchor='center', justify='center')
checking_label.place(relx=0.5, rely=0, relheight=1, anchor="n")
checking_frame.place(relx=0, rely=0, relwidth=1, relheight=1)
# end
# choice frame
choice_frame: ClassVar = Frame(updater_window)
choice_frame.configure(background="#fff")
choice_label: ClassVar = ttk.Label(choice_frame, text="The latest version of Sounder is already installed.\n"
                                                      " Would you like to install it anyway?", font='Bahnschrift 11'
                                   , anchor='center', justify='center')
choice_install_button: ClassVar = ttk.Button(choice_frame, text="INSTALL", cursor="hand2", takefocus=False,
                                             command=update_task)
choice_exit_button: ClassVar = ttk.Button(choice_frame, text="EXIT", cursor="hand2", takefocus=False, command=close)
choice_label.place(relx=0.5, rely=0.1, anchor="n")
choice_install_button.place(relx=0.3, rely=0.6, anchor="n")
choice_exit_button.place(relx=0.7, rely=0.6, anchor="n")
choice_frame.place(relx=0, rely=0, relwidth=1, relheight=1)
# end
# finish frame
finish_frame: ClassVar = Frame(updater_window)
finish_frame.configure(background="#fff")
finish_label: ClassVar = ttk.Label(finish_frame, text="All done!\nLaunching sounder in 3s"
                                   , anchor='center', justify='center', font='Bahnschrift 17')
finish_label.place(relx=0, rely=0, relwidth=1, relheight=1)
finish_frame.place(relx=0, rely=0, relwidth=1, relheight=1)
# end
Thread(target=init, daemon=True).start()

updater_window.mainloop()


src/Elevator.py
try:
    import os
    import threading
    import json
    from tkinter import *
    from tkinter import ttk
    from typing import ClassVar, Dict, List
    import requests
    import zipfile
    import random
    import ctypes
except ImportError:
    sys.exit(2)

# dir
sounder_dir: str = os.getcwd()
# sounder_dir: str = os.path.dirname(sys.executable)
# end
main_window: ClassVar = Tk()
main_window.geometry("300x150")
main_window.title("Elevator")
main_window.iconbitmap(sounder_dir + "\\icon.ico")
main_window.resizable(width=FALSE, height=FALSE)
main_theme = ttk.Style()
main_theme.theme_use('clam')
config: Dict
words: List = ["I swear it's almost done", "Counting to infinity", "Bending the spoon", "Locating the required gigapixels to render", "Spinning up the hamster", "We're building the buildings as fast as we can", "Keep calm and npm install", "Git happens", "Didn't know paint dried so quickly", "Dividing by zero", "Loading funny message", "Leaking into memory"]
server_version: str


def is_admin():
    try:
        return ctypes.windll.shell32.IsUserAnAdmin()
    except:
        return False


def load_config() -> bool:
    global config
    if os.path.isfile('cfg.json'):
        try:
            with open('cfg.json', 'r') as data:
                config = json.load(data)
            return True
        except:
            pass
    return False


def check_for_update() -> bool:
    global config, server_version
    try:
        server_version = requests.get("https://raw.githubusercontent.com/losek1/Sounder3/master/updates/version.txt").text
        if int(config["version"].replace(".", "")) < int(server_version.replace(".", "")):
            return True
        return False
    except:
        return False


def download_package() -> bool:
    global server_version
    try:
        with open("package.zip", 'wb') as file:
            file.write(requests.get("https://github.com/losek1/Sounder3/releases/download/v" + str(server_version) + "/package.zip").content)
        return True
    except:
        return False


def update() -> bool:
    global sounder_dir
    try:
        with zipfile.ZipFile('package.zip', 'r') as zip_ref:
            zip_ref.extractall(sounder_dir)
        os.remove('package.zip')
        update_progress.stop()
    except:
        return False
    return True


def init() -> None:
    if load_config():
        apply_theme()
        if check_for_update():
            if download_package():
                if update():
                    close()


def close():
    if os.path.isfile("Sounder3.exe"):
        os.system("start Sounder3.exe")
    main_window.destroy()
    sys.exit(0)


def apply_theme() -> None:
    global config
    if config["theme"] == "light":
        main_theme.configure("W.TLabel", background='#fff', foreground='#000', border='0')
        main_theme.configure("W.Horizontal.TProgressbar", foreground='#000', background='#000', lightcolor='#fff'
                             , darkcolor='#fff', bordercolor='#fff', troughcolor='#fff')
        main_window.configure(background="#fff")
        update_label.configure(background="#fff", foreground="#000")
        funny_label.configure(background="#fff", foreground="#000")
        info_label.configure(background="#fff", foreground="#000")
    elif config["theme"] == "dark":
        main_theme.configure("W.TLabel", foreground='#fff', background='#000', border='0')
        main_theme.configure("W.Horizontal.TProgressbar", foreground='#000', background='#1e88e5', lightcolor='#000'
                             , darkcolor='#000', bordercolor='#000', troughcolor='#000')
        main_window.configure(background="#000")
        update_label.configure(background="#000", foreground="#fff")
        funny_label.configure(background="#000", foreground="#fff")
        info_label.configure(background="#000", foreground="#fff")


# UI
update_label: ClassVar = ttk.Label(main_window, text="Updating Sounder", font='Bahnschrift 18', anchor="center")
funny_label: ClassVar = ttk.Label(main_window, text=random.SystemRandom().choice(words), font='Bahnschrift 10', anchor="center")
info_label: ClassVar = ttk.Label(main_window, text="This can take a while", font='Bahnschrift 9', anchor="center")
update_progress: ClassVar = ttk.Progressbar(main_window, orient=HORIZONTAL, mode="indeterminate", style="W.Horizontal.TProgressbar")
update_label.place(relx=0, rely=0.02, relwidth=1)
funny_label.place(relx=0, rely=0.3, relwidth=1)
info_label.place(relx=0, rely=0.8, relwidth=1)
update_progress.place(relx=0.5, rely=0.917, relwidth=0.9, height=12, anchor="n")
# end
if is_admin():
    update_progress.start(15)
    init_thread = threading.Thread(target=init, )
    init_thread.daemon = True
    init_thread.start()
else:
    ctypes.windll.shell32.ShellExecuteW(None, "runas", sys.executable, __file__, None, 0)
    sys.exit(1)
main_window.protocol("WM_DELETE_WINDOW", close)
main_window.mainloop()


src/GIFTester.py
from tkinter import *
from PIL import Image, ImageSequence, ImageTk
from threading import Thread
from tkinter import ttk
from typing import ClassVar, List
from time import sleep
main_window: ClassVar = Tk()
main_window.title("ANIMATION")
main_window.geometry("250x250")
main_window.iconbitmap("icon.ico")
# main_window.resizable(width=FALSE, height=FALSE)
main_window.configure(background="#fff")
updater_theme = ttk.Style()
updater_theme.theme_use('clam')
updater_theme.configure("TLabel", background='#fff', foreground='#000', border='0')
img_frames: List = []


def load_img():
    global img_frames
    img: ClassVar = Image.open("download_light.gif")
    for frame in ImageSequence.Iterator(img):
        img_frames.append(ImageTk.PhotoImage(frame.copy().convert('RGBA')))
    while True:
        for frame in img_frames:
            image_label.configure(image=frame)
            sleep(0.015)



image_label: ClassVar = ttk.Label(main_window, image=None)
Thread(target=load_img, daemon=True).start()
image_label.place(relx=0.5, rely=0.5, anchor="center")
main_window.mainloop()

src/Sounder_compiler.py
try:
    import os
    import sys
    import time
    import threading
    import json
    import shutil
    from tkinter import *
    from tkinter import ttk
    import subprocess
    import zipfile
    import logging
    from typing import ClassVar, Dict, List
except ImportError as error_obj:
    print(error_obj)
    sys.exit(1)

# dir
sounder_dir: str = os.getcwd()
# sounder_dir: str = os.path.dirname(sys.executable)
user_path: str = os.path.expanduser("~")
# main window
main_window: ClassVar = Tk()
main_window.geometry("950x600")
main_window.minsize(600, 500)
main_window.configure(background="#fff")
main_window.title("Sounder3 Builder")
main_window.iconbitmap(sounder_dir + "\\icon.ico")
# style
main_theme: ClassVar = ttk.Style()
main_theme.theme_use('clam')
main_theme.configure("TButton", background='#fff', relief="flat", font=('Bahnschrift', 13), foreground='#000')
main_theme.map("TButton", background=[('pressed', '!disabled', '#fff'), ('active', '#eee')])
main_theme.configure("Vertical.TScrollbar", gripcount=0, relief="flat", background="#fff", darkcolor="#fff"
                     , lightcolor="#fff", troughcolor="#fff", bordercolor="#fff", arrowcolor="#000")
main_theme.map("Vertical.TScrollbar", background=[('pressed', '!disabled', '#eee'), ('active', '#eee')])
# end


def py_builder() -> bool:
    global sounder_dir, user_path
    try:
        os.chdir(sounder_dir)
        push_info("\n\nSounder Builder: The building process has started.")
        command: List = ['pyinstaller', '-w', '-i', 'icon.ico', 'Sounder3.py']
        files_to_copy: List = []
        path_to_dist: str = os.path.join(sounder_dir, "dist\\Sounder3")
        path_to_desktop = os.path.join(user_path, "Desktop")
        config = {"refresh_time": 1.0, "theme": "light", "version": "0.0.0", "transition_duration": 1,
                  "gtr_buffer": False, "mode": "r_n",
                  "last_song": "", "continue": False, "path": user_path + "\\Music", "fade": False, "debug": False,
                  "update": True}
    except Exception as error_obj:
        logging.error(error_obj, exc_info=True)
        push_info("Sounder Builder: Error")
        return False
    try:
        shell: ClassVar = subprocess.Popen(command, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
    except subprocess.CalledProcessError as grepexc:
        push_info("\n\n" + str(grepexc.output))
        return False
    for line in iter(shell.stdout.readline, b''):
        push_info(str(line.decode("utf-8")))
    shell.wait()
    try:
        if shell.returncode == 0:
            time.sleep(2)
            push_info("\n\nSounder Builder: Preparing for files transfer.")
            if os.path.isdir("__pycache__"):
                push_info("Sounder Builder: Removing __pycache__")
                # shutil.rmtree("__pycache__")
            if os.path.isdir("build"):
                push_info("Sounder Builder: Removing build.")
                # shutil.rmtree("build")
            if os.path.isfile("Sounder3.spec"):
                push_info("Sounder Builder: Removing Sounder3.spec")
                os.remove("Sounder3.spec")
            if os.path.isdir("dist"):
                for file in os.listdir(sounder_dir):
                    if file.endswith(".png") or file.endswith(".ico") or file.endswith(".txt") or file.endswith(".log") or file.endswith(".json") or file == "Updater.exe":
                        files_to_copy.append(file)
                for file in files_to_copy:
                    push_info("Sounder Builder: Copying " + str(file))
                    shutil.copy(file, path_to_dist)
                push_info("Sounder Builder: Preparing packages.")
                push_info("Sounder Builder: Removing config file.")
                os.remove(os.path.join(path_to_dist, "cfg.json"))
                time.sleep(1)
                push_info("Sounder Builder: Preparing zip.")
                os.chdir(path_to_dist)
                with zipfile.ZipFile(os.path.join(path_to_desktop, "package.zip"), 'w') as zip_obj:
                    for file in os.listdir(path_to_dist):
                        if file == "Updater.exe":
                            continue
                        zip_obj.write(file)
                time.sleep(2)
                push_info("Sounder Builder: Preparing cfg.json")
                with open('cfg.json', 'w') as data:
                    json.dump(config, data)
                os.chdir(path_to_desktop)
                push_info("Sounder Builder: Preparing final release.")
                shutil.move(path_to_dist, path_to_desktop)
                os.system(str("copy " + os.path.join(sounder_dir, "dist") + " " + path_to_desktop + "/o/h/e/k/f/c"))
                push_info("Sounder Builder: Cleaning folders.")
                os.chdir(sounder_dir)
                shutil.rmtree("dist")
                push_info("\n\nSounder Builder: Done.")
    except Exception as error_obj:
        logging.error(error_obj, exc_info=True)
        push_info("Sounder Builder: Error")
        return False
    return True


def build() -> None:
    build_thread: ClassVar = threading.Thread(target=py_builder, )
    build_thread.daemon = True
    build_thread.start()


def push_info(info) -> None:
    build_info.configure(state="normal")
    build_info.insert(END, str(info).rstrip() + "\n")
    build_info.see("end")
    build_info.configure(state="disabled")


info_frame: ClassVar = Frame()
build_scroll_bar: ClassVar = ttk.Scrollbar(info_frame, orient=VERTICAL, cursor="hand2", takefocus=False)
build_info: ClassVar = Text(info_frame, selectbackground="#fff", selectforeground="#000", bd=0, cursor="arrow"
                            , takefocus=0, font=('Bahnschrift', 10), yscrollcommand=build_scroll_bar.set)
build_scroll_bar.configure(command=build_info.yview)
build_info.pack(side=LEFT, fill=BOTH, expand=True)
build_scroll_bar.pack(side=LEFT, fill=Y)

info_frame.place(relx=0.5, rely=0.01, relwidth=0.99, relheight=0.9, anchor="n")
build_button: ClassVar = ttk.Button(main_window, text="Build", takefocus=False, cursor="hand2", command=build)
build_button.place(relx=0.5, rely=0.915, relwidth=0.5, anchor="n")

push_info("Welcome to Sounder3 Builder\nVersion 1.0.0\nby Mateusz Perczak")
logging.basicConfig(filename=sounder_dir + "\\errors.log", level=logging.ERROR)

main_window.mainloop()
